src/crimson/osd/backfill_state.cc

```diff
@@ -225,7 +225,7 @@ bool BackfillState::Enqueuing::should_rescan_primary(
   const BackfillInterval& backfill_info) const
 {
   return backfill_info.begin <= earliest_peer_backfill(peer_backfill_info) &&
-	 !backfill_info.extends_to_end();
+	 !backfill_info.extends_to_end() && backfill_info.empty();
```

Comment:
I'm confused by this condition and the one  in ReplicasScanning::replica_needs_scan.

```backfill_info.begin <= earliest_peer_backfill(peer_backfill_info)``` and the corresponding check in replica_needs_scan ```replica_backfill_info.begin <= local_backfill_info.begin``` seem wrong.  It seems like the check should be whether there's overlap between the interval and the next thing that needs to be backfilled -- something like

```backfill_info.end <= earliest_peer_backfill(peer_backfill_info)``` and ```replica_backfill_info.end <= local_backfill_info.begin```

The empty() check is probably fine, but I'm not sure if it's actually necessary.  For there to be no overlap, the contents of the previous scan would have to already have been backfilled and removed.

Reply:
> backfill_info.begin <= earliest_peer_backfill(peer_backfill_info) and the corresponding check in replica_needs_scan replica_backfill_info.begin <= local_backfill_info.begin seem wrong. It seems like the check should be whether there's overlap between the interval and the next thing that needs to be backfilled -- something like

I think those checks are copied from `PrimaryLogPG::recover_backfill()`

https://github.com/ceph/ceph/blob/d7a74a60764825edeb5e7d99435737917447fd1a/src/osd/PrimaryLogPG.cc#L13946-L13947
https://github.com/ceph/ceph/blob/d7a74a60764825edeb5e7d99435737917447fd1a/src/osd/PrimaryLogPG.cc#L13965-L13966

Reply:
Ah, I see now.  If ```backfill_info.begin <= earliest_peer_backfill(peer_backfill_info)``` and it's empty (and doesn't extend to end), they don't actually overlap and we need to rescan primary to get more entries.  Checking end could also work, but this change to match PrimaryLogPG is fine.